src/node_ffi.cc

```diff
@@ -0,0 +1,381 @@
+
+void FFIDefinition::readValue(const int i,
+                              const Local<Value> input,
+                              ffi_raw* output) const {
+  switch (types[i]->type) {
+    case FFI_TYPE_VOID:
+      output->uint = 0;
+      break;
+    case FFI_TYPE_UINT8:
+    case FFI_TYPE_UINT16:
+    case FFI_TYPE_UINT32:
+    case FFI_TYPE_UINT64:
+      output->uint = readUInt64(input);
+      break;
+    case FFI_TYPE_SINT8:
+    case FFI_TYPE_SINT16:
+    case FFI_TYPE_SINT32:
+    case FFI_TYPE_SINT64:
+      output->sint = readInt64(input);
+      break;
+    case FFI_TYPE_FLOAT:
+      output->flt = (float)readDouble(input);
+      break;
+    case FFI_TYPE_POINTER:
+      output->ptr = readAddress(input);
+      break;
+    default:
+      UNREACHABLE("Bad FFI type");
+  }
+}
+
+Local<Value> FFIDefinition::wrapValue(const int i,
+                                      Isolate* isolate,
+                                      const ffi_raw* input) const {
+  switch (types[i]->type) {
+    case FFI_TYPE_VOID:
+      return Undefined(isolate);
+    case FFI_TYPE_UINT8:
+    case FFI_TYPE_UINT16:
+    case FFI_TYPE_UINT32:
+      return Uint32::NewFromUnsigned(isolate, (uint32_t)input->uint);
+    case FFI_TYPE_UINT64:
+      return BigInt::NewFromUnsigned(isolate, input->uint);
+    case FFI_TYPE_SINT8:
+    case FFI_TYPE_SINT16:
+    case FFI_TYPE_SINT32:
+      return Int32::New(isolate, (int32_t)input->sint);
+    case FFI_TYPE_SINT64:
+      return BigInt::New(isolate, input->sint);
+    case FFI_TYPE_FLOAT:
+      return Number::New(isolate, input->flt);
+    case FFI_TYPE_POINTER:
+      return External::New(isolate, input->ptr);
+    default:
+      UNREACHABLE("Bad FFI type");
+  }
+}
+
+FFIInvoker::FFIInvoker(const char* defStr, const void* address)
+    : FFIDefinition(defStr) {
+  invoker = FFI_FN(address);
+  datas = std::make_unique<ffi_raw[]>(cif.nargs);
+}
+
+void FFIInvoker::setParam(const int i, const Local<Value> value) {
+  readValue(i, value, &datas[i - 1]);
+}
+
+Local<Value> FFIInvoker::doInvoke(Isolate* isolate) {
+  ffi_raw result;
+  ffi_raw_call(&cif, invoker, &result, datas.get());
+  return wrapValue(0, isolate, &result);
+}
+
+FFICallback::FFICallback(const char* defStr) : FFIDefinition(defStr) {
+  frc = (ffi_raw_closure*)ffi_closure_alloc(RCS, &address);
+  if (!frc) {
+    UNREACHABLE("ffi_closure_alloc Failed");
+  }
+  if (ffi_prep_raw_closure(frc, &cif, ffiCallback, this) != FFI_OK) {
+    UNREACHABLE("ffi_prep_raw_closure Failed");
+  }
+}
+
+void FFICallback::setCallback(Isolate* isolate, const Local<Value> value) {
+  if (value->IsFunction()) {
+    const auto function = value.As<Function>();
+    callback.Reset(isolate, function);
+  }
+}
+
+void FFICallback::doCallback(ffi_raw* result,
+                             const int argc,
+                             const ffi_raw* args) const {
+  const auto isolate = Isolate::GetCurrent();
+  const auto params = std::make_unique<Local<Value>[]>(argc);
```

Comment:
v8::Local<T> cannot be safely allocated on the heap like this. You'll want to use v8::LocalVector instead.